File: parser/parser_test.go
package parser

import (
	"testing"

	"github.com/kawaz/mermaid-aa/types"
)

func TestParseDirection(t *testing.T) {
	tests := []struct {
		input    string
		expected types.Direction
	}{
		{"graph TD; A-->B", types.TD},
		{"graph TB; A-->B", types.TB},
		{"graph LR; A-->B", types.LR},
		{"graph RL; A-->B", types.RL},
		{"graph BT; A-->B", types.BT},
		{"flowchart TD; A-->B", types.TD},
		{"graph; A-->B", types.TB}, // default
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			fc, err := ParseFlowchart(tt.input)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if fc.Direction != tt.expected {
				t.Errorf("direction = %v, want %v", fc.Direction, tt.expected)
			}
		})
	}
}

func TestParseNodes(t *testing.T) {
	tests := []struct {
		input  string
		nodeID string
		label  string
		shape  types.NodeShape
	}{
		{"graph TD; A[Hello]", "A", "Hello", types.Rectangle},
		{"graph TD; B(World)", "B", "World", types.Rounded},
		{"graph TD; C([Stadium])", "C", "Stadium", types.Stadium},
		{"graph TD; D{Diamond}", "D", "Diamond", types.Diamond},
		{"graph TD; E{{Hexagon}}", "E", "Hexagon", types.Hexagon},
		{"graph TD; F((Circle))", "F", "Circle", types.Circle},
		{"graph TD; G[/Parallel/]", "G", "Parallel", types.Parallelogram},
		{`graph TD; H[/Trap\]`, "H", "Trap", types.Trapezoid},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			fc, err := ParseFlowchart(tt.input)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			node, ok := fc.Nodes[tt.nodeID]
			if !ok {
				t.Fatalf("node %s not found", tt.nodeID)
			}
			if node.Label != tt.label {
				t.Errorf("label = %q, want %q", node.Label, tt.label)
			}
			if node.Shape != tt.shape {
				t.Errorf("shape = %v, want %v", node.Shape, tt.shape)
			}
		})
	}
}

func TestParseEdges(t *testing.T) {
	tests := []struct {
		input  string
		fromID string
		toID   string
		style  types.EdgeStyle
	}{
		{"graph TD; A-->B", "A", "B", types.SolidArrow},
		{"graph TD; A-.->B", "A", "B", types.DottedArrow},
		{"graph TD; A==>B", "A", "B", types.ThickArrow},
		{"graph TD; A---B", "A", "B", types.Open},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			fc, err := ParseFlowchart(tt.input)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if len(fc.Edges) == 0 {
				t.Fatal("no edges found")
			}
			edge := fc.Edges[0]
			if edge.From != tt.fromID {
				t.Errorf("from = %q, want %q", edge.From, tt.fromID)
			}
			if edge.To != tt.toID {
				t.Errorf("to = %q, want %q", edge.To, tt.toID)
			}
			if edge.Style != tt.style {
				t.Errorf("style = %v, want %v", edge.Style, tt.style)
			}
		})
	}
}

func TestParseChain(t *testing.T) {
	input := "graph TD; A-->B-->C"
	fc, err := ParseFlowchart(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(fc.Nodes) != 3 {
		t.Errorf("nodes count = %d, want 3", len(fc.Nodes))
	}
	if len(fc.Edges) != 2 {
		t.Errorf("edges count = %d, want 2", len(fc.Edges))
	}

	// Check edges
	if fc.Edges[0].From != "A" || fc.Edges[0].To != "B" {
		t.Errorf("edge 0 = %s-->%s, want A-->B", fc.Edges[0].From, fc.Edges[0].To)
	}
	if fc.Edges[1].From != "B" || fc.Edges[1].To != "C" {
		t.Errorf("edge 1 = %s-->%s, want B-->C", fc.Edges[1].From, fc.Edges[1].To)
	}
}

func TestParseWithLabels(t *testing.T) {
	input := "graph TD; A[Start]-->B[End]"
	fc, err := ParseFlowchart(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if fc.Nodes["A"].Label != "Start" {
		t.Errorf("A label = %q, want %q", fc.Nodes["A"].Label, "Start")
	}
	if fc.Nodes["B"].Label != "End" {
		t.Errorf("B label = %q, want %q", fc.Nodes["B"].Label, "End")
	}
}

func TestParseMultiLine(t *testing.T) {
	input := `graph TD
    A[Start]
    B[Process]
    C[End]
    A-->B
    B-->C`

	fc, err := ParseFlowchart(input)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(fc.Nodes) != 3 {
		t.Errorf("nodes count = %d, want 3", len(fc.Nodes))
	}
	if len(fc.Edges) != 2 {
		t.Errorf("edges count = %d, want 2", len(fc.Edges))
	}
}

File: parser_test.ts
/**
 * Tests for Mermaid parser
 */

import { assertEquals } from "@std/assert";
import { parse } from "./parser.ts";

Deno.test("parse - Simple flowchart with direction", () => {
  const input = `flowchart TD
    A --> B`;

  const result = parse(input);

  assertEquals(result.direction, "TB");
  assertEquals(result.nodes.size, 2);
  assertEquals(result.edges.length, 1);
  assertEquals(result.edges[0].from, "A");
  assertEquals(result.edges[0].to, "B");
});

Deno.test("parse - Graph keyword works like flowchart", () => {
  const input = `graph LR
    A --> B`;

  const result = parse(input);

  assertEquals(result.direction, "LR");
  assertEquals(result.nodes.size, 2);
});

Deno.test("parse - Rectangle node", () => {
  const input = `flowchart TD
    A[Hello World]`;

  const result = parse(input);

  assertEquals(result.nodes.size, 1);
  const node = result.nodes.get("A");
  assertEquals(node?.shape, "rectangle");
  assertEquals(node?.label, "Hello World");
});

Deno.test("parse - Rounded node", () => {
  const input = `flowchart TD
    A(Hello)`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "rounded");
  assertEquals(node?.label, "Hello");
});

Deno.test("parse - Stadium node", () => {
  const input = `flowchart TD
    A([Stadium])`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "stadium");
  assertEquals(node?.label, "Stadium");
});

Deno.test("parse - Diamond node", () => {
  const input = `flowchart TD
    A{Decision}`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "diamond");
  assertEquals(node?.label, "Decision");
});

Deno.test("parse - Hexagon node", () => {
  const input = `flowchart TD
    A{{Hexagon}}`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "hexagon");
  assertEquals(node?.label, "Hexagon");
});

Deno.test("parse - Circle node", () => {
  const input = `flowchart TD
    A((Circle))`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "circle");
  assertEquals(node?.label, "Circle");
});

Deno.test("parse - Parallelogram node", () => {
  const input = `flowchart TD
    A[/Parallelogram/]`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "parallelogram");
  assertEquals(node?.label, "Parallelogram");
});

Deno.test("parse - Trapezoid node", () => {
  const input = String.raw`flowchart TD
    A[/Trapezoid\]`;

  const result = parse(input);

  const node = result.nodes.get("A");
  assertEquals(node?.shape, "trapezoid");
  assertEquals(node?.label, "Trapezoid");
});

Deno.test("parse - Solid arrow", () => {
  const input = `flowchart TD
    A --> B`;

  const result = parse(input);

  assertEquals(result.edges[0].style, "solid");
});

Deno.test("parse - Dotted arrow", () => {
  const input = `flowchart TD
    A -.-> B`;

  const result = parse(input);

  assertEquals(result.edges[0].style, "dotted");
});

Deno.test("parse - Thick arrow", () => {
  const input = `flowchart TD
    A ==> B`;

  const result = parse(input);

  assertEquals(result.edges[0].style, "thick");
});

Deno.test("parse - Open connection", () => {
  const input = `flowchart TD
    A --- B`;

  const result = parse(input);

  assertEquals(result.edges[0].style, "open");
});

Deno.test("parse - Arrow with label", () => {
  const input = `flowchart TD
    A --|label|--> B`;

  const result = parse(input);

  assertEquals(result.edges[0].label, "label");
});

Deno.test("parse - Multiple statements on one line", () => {
  const input = `flowchart TD
    A --> B; B --> C`;

  const result = parse(input);

  assertEquals(result.edges.length, 2);
  assertEquals(result.edges[0].from, "A");
  assertEquals(result.edges[0].to, "B");
  assertEquals(result.edges[1].from, "B");
  assertEquals(result.edges[1].to, "C");
});

Deno.test("parse - Different directions", () => {
  assertEquals(parse("flowchart TB\n A-->B").direction, "TB");
  assertEquals(parse("flowchart TD\n A-->B").direction, "TB");
  assertEquals(parse("flowchart BT\n A-->B").direction, "BT");
  assertEquals(parse("flowchart LR\n A-->B").direction, "LR");
  assertEquals(parse("flowchart RL\n A-->B").direction, "RL");
});

Deno.test("parse - Comments are ignored", () => {
  const input = `flowchart TD
    %% This is a comment
    A --> B`;

  const result = parse(input);

  assertEquals(result.nodes.size, 2);
  assertEquals(result.edges.length, 1);
});

Deno.test("parse - CJK labels", () => {
  const input = `flowchart TD
    A[日本語] --> B[中国語]`;

  const result = parse(input);

  assertEquals(result.nodes.get("A")?.label, "日本語");
  assertEquals(result.nodes.get("B")?.label, "中国語");
});

Deno.test("parse - Default direction is TB", () => {
  const input = `flowchart
    A --> B`;

  const result = parse(input);

  assertEquals(result.direction, "TB");
});
